From 7e7c1260898355937b73457edd00a0c73e47c81b Mon Sep 17 00:00:00 2001
From: Christian Ehrlicher <ch.ehrlicher@gmx.de>
Date: Mon, 27 Oct 2025 20:52:39 +0100
Subject: [PATCH 537/553] Windows11Style: draw pushbutton menu in
 CE_PushButtonBevel

Draw the menu indicator in CE_PushButtonBevel instead CE_PushButtonLabel
to be in sync with other styles and make it work with QStyleSheetStyle.

Fixes: QTBUG-141499
Change-Id: I4d7d36191bcc752935582ba229822bd854f85624
Reviewed-by: Wladimir Leuschner <wladimir.leuschner@qt.io>
(cherry picked from commit 133ec7d3721ac87f3186096cda0fdcb1999829b3)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 .../styles/modernwindows/qwindows11style.cpp  | 45 ++++++++++---------
 1 file changed, 25 insertions(+), 20 deletions(-)

diff --git a/src/plugins/styles/modernwindows/qwindows11style.cpp b/src/plugins/styles/modernwindows/qwindows11style.cpp
index 6d7e691c41c..50db65ef21e 100644
--- a/src/plugins/styles/modernwindows/qwindows11style.cpp
+++ b/src/plugins/styles/modernwindows/qwindows11style.cpp
@@ -1423,26 +1423,6 @@ void QWindows11Style::drawControl(ControlElement element, const QStyleOption *op
             if (!proxy()->styleHint(SH_UnderlineShortcut, btn, widget))
                 tf |= Qt::TextHideMnemonic;
 
-            if (btn->features & QStyleOptionButton::HasMenu) {
-                QPainterStateGuard psg(painter);
-
-                const auto indSize = proxy()->pixelMetric(PM_MenuButtonIndicator, btn, widget);
-                const auto indRect = QRect(btn->rect.right() - indSize - contentItemHMargin, textRect.top(),
-                                           indSize + contentItemHMargin, btn->rect.height());
-                const auto vindRect = visualRect(btn->direction, btn->rect, indRect);
-                textRect.setWidth(textRect.width() - indSize);
-
-                int fontSize = painter->font().pointSize();
-                QFont f(d->assetFont);
-                f.setPointSize(qRound(fontSize * 0.9f)); // a little bit smaller
-                painter->setFont(f);
-                QColor penColor = option->palette.color(isEnabled ? QPalette::Active : QPalette::Disabled,
-                                                        QPalette::Text);
-                if (isEnabled)
-                    penColor.setAlpha(percentToAlpha(60.63)); // fillColorTextSecondary
-                painter->setPen(penColor);
-                painter->drawText(vindRect, Qt::AlignCenter, ChevronDownMed);
-            }
             if (!btn->icon.isNull()) {
                 //Center both icon and text
                 QIcon::Mode mode = isEnabled ? QIcon::Normal : QIcon::Disabled;
@@ -1469,6 +1449,8 @@ void QWindows11Style::drawControl(ControlElement element, const QStyleOption *op
         break;
     case CE_PushButtonBevel:
         if (const QStyleOptionButton *btn = qstyleoption_cast<const QStyleOptionButton *>(option))  {
+            using namespace StyleOptionHelper;
+
             QRectF rect = btn->rect.marginsRemoved(QMargins(2, 2, 2, 2));
             painter->setPen(Qt::NoPen);
             if (btn->features.testFlag(QStyleOptionButton::Flat)) {
@@ -1495,6 +1477,29 @@ void QWindows11Style::drawControl(ControlElement element, const QStyleOption *op
                 painter->setPen(defaultButton ? WINUI3Colors[colorSchemeIndex][controlStrokeOnAccentSecondary]
                                               : WINUI3Colors[colorSchemeIndex][controlStrokeSecondary]);
             }
+            if (btn->features.testFlag(QStyleOptionButton::HasMenu)) {
+                QPainterStateGuard psg(painter);
+
+                const bool isEnabled = !isDisabled(option);
+                QRect textRect = btn->rect.marginsRemoved(QMargins(contentHMargin, 0, contentHMargin, 0));
+                const auto indSize = proxy()->pixelMetric(PM_MenuButtonIndicator, btn, widget);
+                const auto indRect =
+                        QRect(btn->rect.right() - indSize - contentItemHMargin, textRect.top(),
+                              indSize + contentItemHMargin, btn->rect.height());
+                const auto vindRect = visualRect(btn->direction, btn->rect, indRect);
+                textRect.setWidth(textRect.width() - indSize);
+
+                int fontSize = painter->font().pointSize();
+                QFont f(d->assetFont);
+                f.setPointSize(qRound(fontSize * 0.9f)); // a little bit smaller
+                painter->setFont(f);
+                QColor penColor = option->palette.color(
+                        isEnabled ? QPalette::Active : QPalette::Disabled, QPalette::Text);
+                if (isEnabled)
+                    penColor.setAlpha(percentToAlpha(60.63)); // fillColorTextSecondary
+                painter->setPen(penColor);
+                painter->drawText(vindRect, Qt::AlignCenter, ChevronDownMed);
+            }
         }
         break;
     case CE_MenuBarItem:
-- 
2.51.2

